### PR TITLE
Module4 tasks

### PR DESCRIPTION
**Pull request by Nastya-Korsak**
Task1 and Task2

### CODE REVIEW

File: Module_4/FileSystemVisitorLibrary.Tests/UnitTest1.cs [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+                else
+                {
+                    return Actions.Continue;
+                }
+            };
+
+            var items = visitor.Search(@"d:\test1");
+
+            // Assert
+            items.Should().BeEquivalentTo(@"d:\test1\tes1_1", @"d:\test1\tes1_2");
```
Comment (alexander-trishin) [inline Module_4/FileSystemVisitorLibrary.Tests/UnitTest1.cs:154]: Here is a bug in the `FileSystemVisitor` class. `FilteredFileFound` should be fired only for files with `.txt` extension. So the final result should be:
```
d:\test1\im.gif
d:\test1\test1_1
d:\test1\test1_2
```

File: Module_4/FileSystemVisitorLibrary.Tests/UnitTest1.cs [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+                }
+            };
+
+            foreach (var f in visitor.Search(@"d:\test1"))
+            {
+                items.Add(f);
+            }
+
+            // Assert
+            items.Select(i => i.Substring(i.LastIndexOf('\\'))).Should().NotContain("test");
```
Comment (alexander-trishin) [inline Module_4/FileSystemVisitorLibrary.Tests/UnitTest1.cs:198]: You're reinventing the wheel here 😃 

```suggestion
            items.Select(info => new DirectoryInfo(info).Name).Should().NotContain("test");
```

File: Module_4/FileSystemVisitorLibrary.Tests/UnitTest1.cs [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+            items.Last().Should().Be(@"d:\test1\test1_1\dir3");
+        }
+
+        [Fact]
+        public void Search_NullPath_ArgumentNullException()
```
Comment (alexander-trishin) [inline Module_4/FileSystemVisitorLibrary.Tests/UnitTest1.cs:203]:
```suggestion
        public void Search_PathParameterIsNull_ShouldThrowArgumentNullException()
```

File: Module_4/FileSystemVisitorLibrary.Tests/UnitTest1.cs [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+            });
+
+            // Act
+            var visitor = new FileSystemVisitor(fileSystem);
+            Func<IEnumerable<string>> searchFunc = () => visitor.Search(null);
+
+            // Assert
+            searchFunc.Enumerating().Should().Throw<ArgumentNullException>()
+                .WithMessage("Directory path is null")
+                .And.ParamName.Should().Be(null);
```
Comment (alexander-trishin) [inline Module_4/FileSystemVisitorLibrary.Tests/UnitTest1.cs:220]: `ParamName` should never be `null`. It's a bug in `FileSystemVisitor`.

File: Module_4/FileSystemVisitorLibrary/Actions.cs [hunk 5]
```diff
@@ -0,0 +1,9 @@
+﻿namespace FileSystemVisitorLibrary
+{
+    public enum Actions
```
Comment (alexander-trishin) [inline Module_4/FileSystemVisitorLibrary/Actions.cs:3]: Please use more unique names for top-level elements like classes, interfaces and enums.

File: Module_4/FileSystemVisitorLibrary/FileSystemVisitor.cs [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+using System.Collections.Generic;
+using System.IO;
+using System.IO.Abstractions;
+using System.Linq;
+
+namespace FileSystemVisitorLibrary
+{
+    public class FileSystemVisitor
+    {
+        private readonly Func<FileSystemInfo, bool> _filterFunc;
```
Comment (alexander-trishin) [inline Module_4/FileSystemVisitorLibrary/FileSystemVisitor.cs:11]: It's clear that this is a `Func` when we look at the type of the variable.

```suggestion
        private readonly Func<FileSystemInfo, bool> _filter;
```

File: Module_4/FileSystemVisitorLibrary/FileSystemVisitor.cs [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+
+        private readonly IFileSystem _fileSystem;
+
+        private Actions? _action = Actions.Continue;
+
+        public FileSystemVisitor(IFileSystem fileSystem, Func<FileSystemInfo, bool> filterFunc = null)
+        {
+            _fileSystem = fileSystem;
```
Comment (alexander-trishin) [inline Module_4/FileSystemVisitorLibrary/FileSystemVisitor.cs:19]: File system is a critical dependency for this class. Incorrect behavior should be prevented as soon as possible.
```suggestion
            _fileSystem = fileSystem ?? throw new ArgumentNullException(nameof(fileSystem));
```

File: Module_4/FileSystemVisitorLibrary/FileSystemVisitor.cs [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+            _filterFunc = filterFunc;
+        }
+
+        public event Action Started;
+
+        public event Action Complited;
```
Comment (alexander-trishin) [inline Module_4/FileSystemVisitorLibrary/FileSystemVisitor.cs:25]: A typo.

```suggestion
        public event Action Completed;
```

File: Module_4/FileSystemVisitorLibrary/FileSystemVisitor.cs [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+
+        public event Func<FileSystemInfo, Actions> FileFound;
```
Comment (alexander-trishin) [inline Module_4/FileSystemVisitorLibrary/FileSystemVisitor.cs:27]: Normally event handlers should match `EventHandler` syntax. [MSDN Guide for events](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/events/how-to-publish-events-that-conform-to-net-framework-guidelines).

File: Module_4/FileSystemVisitorLibrary/FileSystemVisitor.cs [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+
+        public event Func<FileSystemInfo, Actions> FilteredFileFound;
+
+        public event Func<FileSystemInfo, Actions> FilteredDirectoryFound;
+
+        public IEnumerable<string> Search(string path)
+        {
+            if (path == null)
+            {
+                throw new ArgumentNullException(path, "Directory path is null");
```
Comment (alexander-trishin) [inline Module_4/FileSystemVisitorLibrary/FileSystemVisitor.cs:39]: Should be:
```suggestion
                throw new ArgumentNullException(nameof(path));
```

File: Module_4/FileSystemVisitorLibrary/FileSystemVisitor.cs [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+            }
+
+            if (path == string.Empty)
+            {
+                throw new ArgumentException("Directory path is empty", path);
```
Comment (alexander-trishin) [inline Module_4/FileSystemVisitorLibrary/FileSystemVisitor.cs:44]: Please, do not forget to use `nameof` keyword.

```suggestion
                throw new ArgumentException("Directory path is empty", nameof(path));
```

File: Module_4/FileSystemVisitorLibrary/FileSystemVisitor.cs [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+            }
+
+            if (!_fileSystem.Directory.Exists(path))
+            {
+                throw new ArgumentException("Selected directory is not exist", path);
```
Comment (alexander-trishin) [inline Module_4/FileSystemVisitorLibrary/FileSystemVisitor.cs:49]: Try to always use more specific types of exceptions:

```suggestion
                throw new DirectoryNotFoundException($"Selected directory is not exist: {path}");
```

File: Module_4/FileSystemVisitorLibrary/FileSystemVisitor.cs [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+            }
+
+            Started?.Invoke();
+
+            foreach (var f in FileSearch(path))
+            {
+                yield return f;
+            }
+
+            foreach (var d in DirectionSearch(path))
```
Comment (alexander-trishin) [inline Module_4/FileSystemVisitorLibrary/FileSystemVisitor.cs:59]: The recursive algorithm is more applicable for this task. It should also fix the bug where directories are listed only after files.

File: Module_4/FileSystemVisitorLibrary/FileSystemVisitor.cs [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+            {
+                yield return d;
+            }
+
+            Complited?.Invoke();
+        }
+
+        private IEnumerable<string> FileSearch(string path)
```
Comment (alexander-trishin) [inline Module_4/FileSystemVisitorLibrary/FileSystemVisitor.cs:67]: I would recommend to work with `FileSystemInfo` class.

```suggestion
        private IEnumerable<string> FileSearch(FileSystemInfo fileInfo)
```

File: Module_4/FileSystemVisitorLibrary/FileSystemVisitor.cs [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+        {
+            if (_action == Actions.Stop)
+            {
+                yield break;
+            }
+
+            foreach (var f in _fileSystem.Directory.EnumerateFiles(path).Where(f => _filterFunc == null || !_filterFunc(new FileInfo(f))))
```
Comment (alexander-trishin) [inline Module_4/FileSystemVisitorLibrary/FileSystemVisitor.cs:74]: Short names are only applicable in rare cases.

For example: `directories.Select(x => x.Name)` The lambda is pretty small and still readable.

As the method body increases, short names become unreadable.

File: Module_4/FileSystemVisitorLibrary/FileSystemVisitor.cs [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+            {
+                var fileInfo = new FileInfo(f);
+
+                if (_filterFunc == null && FileFound != null)
+                {
+                    _action = FileFound.Invoke(fileInfo);
+                }
+                else if (FilteredFileFound != null)
+                {
+                    _action = FilteredFileFound.Invoke(fileInfo);
```
Comment (alexander-trishin) [inline Module_4/FileSystemVisitorLibrary/FileSystemVisitor.cs:84]: [Bug] It will mark normal files as filtered (because of rows #74 and #82)

**Review by alexander-trishin — CHANGES_REQUESTED** [review]
Hi Nastya,

Unfortunately, I found a few bugs. Please change the code according to my comments.

File: Module_4/FileSystemVisitorLibrary.Tests/UnitTest1.cs [hunk 17]
```diff
@@ -0,0 +1,267 @@
...
+using System.Collections.Generic;
+using System.IO;
+using System.IO.Abstractions.TestingHelpers;
+using System.Linq;
+using FluentAssertions;
+using Xunit;
+
+namespace FileSystemVisitorLibrary.Tests
+{
+    public class UnitTest1
```
[the remark this answers is not in the document]
Comment (Nastya-Korsak) [reply Module_4/FileSystemVisitorLibrary.Tests/UnitTest1.cs]: fixed

File: Module_4/Task1/Task1.sln [hunk 18]
```diff
@@ -0,0 +1,25 @@
+﻿
```
[the remark this answers is not in the document]
Comment (Nastya-Korsak) [reply Module_4/Task1/Task1.sln]: fixed

File: Module_4/FileSystemVisitorLibrary.Tests/UnitTest1.cs [hunk 19]
```diff
@@ -0,0 +1,267 @@
...
+                { @"d:\test1\tes1_2\text2.txt", new MockFileData("Text2") },
+                { @"d:\test1\im.gif", new MockFileData(System.Array.Empty<byte>()) }
+            });
+
+            // Act
+            var visitor = new FileSystemVisitor(fileSystem);
+            var items = visitor.Search(@"d:\test1");
+
+            // Assert
+            items.Should().BeEquivalentTo(@"d:\test1\im.gif", @"d:\test1\tes1_1\text.txt", @"d:\test1\tes1_2\text2.txt", @"d:\test1\tes1_1", @"d:\test1\tes1_2");
```
[the remark this answers is not in the document]
Comment (Nastya-Korsak) [reply Module_4/FileSystemVisitorLibrary.Tests/UnitTest1.cs]: fixed

File: Module_4/FileSystemVisitorLibrary.Tests/UnitTest1.cs [hunk 20]
```diff
@@ -0,0 +1,267 @@
...
+        }
+
+        [Fact]
+        public void Search_FilterTxtFiles_ShouldNotReturnTxtFiles()
+        {
+            // Arrange
+            var fileSystem = new MockFileSystem(new Dictionary<string, MockFileData>
```
[the remark this answers is not in the document]
Comment (Nastya-Korsak) [reply Module_4/FileSystemVisitorLibrary.Tests/UnitTest1.cs]: fixed

File: Module_4/FileSystemVisitorLibrary.Tests/UnitTest1.cs [hunk 21]
```diff
@@ -0,0 +1,267 @@
...
+                { @"d:\test1\tes1_2\text2.txt", new MockFileData("Text2") },
+                { @"d:\test1\im.gif", new MockFileData(System.Array.Empty<byte>()) }
+            });
+
+            // Act
+            var visitor = new FileSystemVisitor(fileSystem);
+
+            visitor.DirectoryFound += (FileSystemInfo e) =>
+            {
+                if (e.Name.Contains("tes1_1"))
```
[the remark this answers is not in the document]
Comment (Nastya-Korsak) [reply Module_4/FileSystemVisitorLibrary.Tests/UnitTest1.cs]: fixed

File: Module_4/FileSystemVisitorLibrary/FileSystemVisitor.cs (hunk 12 above)
Comment (Nastya-Korsak) [reply Module_4/FileSystemVisitorLibrary/FileSystemVisitor.cs]: fixed

File: Module_4/FileSystemVisitorLibrary/FileSystemVisitor.cs (hunk 13 above)
Comment (Nastya-Korsak) [reply Module_4/FileSystemVisitorLibrary/FileSystemVisitor.cs]: fixed

File: Module_4/FileSystemVisitorLibrary/FileSystemVisitor.cs (hunk 14 above)
Comment (Nastya-Korsak) [reply Module_4/FileSystemVisitorLibrary/FileSystemVisitor.cs]: fixed

File: Module_4/FileSystemVisitorLibrary/FileSystemVisitor.cs (hunk 15 above)
Comment (Nastya-Korsak) [reply Module_4/FileSystemVisitorLibrary/FileSystemVisitor.cs]: fixed

File: Module_4/FileSystemVisitorLibrary/FileSystemVisitor.cs (hunk 16 above)
Comment (Nastya-Korsak) [reply Module_4/FileSystemVisitorLibrary/FileSystemVisitor.cs]: fixed

**Review by alexander-trishin — APPROVED** [review]
I see a few typos, but in general: good job! 😄

**Pull request merged**